internal/dataplane/kongstate/kongstate.go

```diff
@@ -141,25 +142,30 @@ func (ks *KongState) FillConsumersAndCredentials(
 				}
 				credConfig[k] = string(v)
 			}
-			credType, ok := credConfig["kongCredType"].(string)
-			if !ok {
-				pushCredentialResourceFailures(
-					fmt.Sprintf("failed to provision credential: invalid kongCredType: type '%T' not string", credType),
-				)
-				continue
+			// try the label first. if it's present, no need to check the field
+			credType, labelOk := secret.Labels[labels.LabelPrefix+labels.CredentialKey]
+			var fieldOk bool
+			if !labelOk {
+				// if no label, fall back to the deprecated field
+				credType, fieldOk = credConfig["kongCredType"].(string)
+				if !fieldOk {
+					pushCredentialResourceFailures(
+						fmt.Sprintf("failed to provision credential: invalid kongCredType: type '%T' not string", credType),
+					)
+					continue
+				}
+			}
+			if fieldOk && !labelOk {
+				failuresCollector.PushResourceFailure("credential only has deprecated kongCredType field, needs "+
```

Comment:
I'm not sure if we should communicate deprecation warnings via resource failures 🤔 I have a feeling they should be reserved for stuff that really might break something because we had to skip the resource, etc. I'd rather suggest adding the deprecation warning to the admission webhook (we can allow the resource, but still return a message with a deprecation warning).

Reply:
We want something that will trigger for existing unmodified resources, whereas the admission webhook will only run when a resource changes.

I agree that the event type is a bit of a misnomer here ("problem" instead of "failure" would cover a broader set of issues), but events in general provide characteristics (applies to existing resources and generates something you can search via kubectl) that are useful here.

AFAIK alternatives here are:

- Only log these resources. This isn't helpful for kubectl searching, but we do have an external script to find the resources. IDK if events are more visible (users could easily not go looking for either).
- Create an event type specifically for deprecations. I didn't think this was worth it for just one resource when we had something that was functionally equivalent already, even if you need to stretch the definition of "failure".

Reply:
> Only log these resources. This isn't helpful for kubectl searching, but we do have an external script to find the resources. IDK if events are more visible (users could easily not go looking for either).

👍 IMO log would be enough for the deprecation warning. We'll have that mentioned in the release notes as well + in the admission webhook.

> Create an event type specifically for deprecations. I didn't think this was worth it for just one resource when we had something that was functionally equivalent already, even if you need to stretch the definition of "failure".

I think that we should keep `KongConfigurationTranslationFailed` events only for cases where the configuration is seriously affected in some way like it's described in the [docs](https://docs.konghq.com/kubernetes-ingress-controller/latest/references/kubernetes-events/). If someone had alerting hooked up for these events they probably wouldn't like to be woken up because of a deprecation warning :)